libs/@guardian/identity-auth/src/autoRenew.ts

```diff
@@ -0,0 +1,146 @@
+import { getCookie } from '@guardian/libs';
+import type { IdentityAuthOptions, IdentityAuthState } from './@types/OAuth';
+import type { AuthStateManager } from './authState';
+import type { Emitter } from './emitter';
+
+/**
+ * @class AutoRenewService
+ * @description Managing the auto renewal of tokens, if enabled
+ */
+export class AutoRenewService {
+	#options: Required<IdentityAuthOptions>;
+	#emitter: Emitter;
+	#authStateManager: AuthStateManager;
+	started = false;
+
+	#renewAtTimeoutId: number | undefined;
+
+	constructor(
+		options: Required<IdentityAuthOptions>,
+		emitter: Emitter,
+		authStateManager: AuthStateManager,
+	) {
+		this.#options = options;
+		this.#emitter = emitter;
+		this.#authStateManager = authStateManager;
+
+		// if there is a storage event, clear the timeout, and start a new one
+		this.#emitter.on('storage', () => {
+			this.#clearRenewAtTimeout();
+			this.#handleAuthStateChange();
+		});
+	}
+
+	/**
+	 * @name clearRenewAtTimeout
+	 * @description Clears the timeout for renewing the access token, if it exists
+	 */
+	#clearRenewAtTimeout() {
+		if (this.#renewAtTimeoutId !== undefined) {
+			window.clearTimeout(this.#renewAtTimeoutId);
+			this.#renewAtTimeoutId = undefined;
+		}
+	}
+
+	/**
+	 * @name setRenewAtTimeout
+	 * @description Sets the timeout for renewing the access token based on the expiry time and the grace period
+	 * @param expiresAt The time at which the access token expires
+	 * @returns The timeout id
+	 */
+	#setRenewAtTimeout(expiresAt: number) {
+		// get the time now
+		const now = Date.now();
+
+		// calculate the time at which the access token should be renewed
+		const renewAt = (expiresAt - this.#options.renewGracePeriod) * 1000;
+
+		// calculate the timeout duration
+		const timeout = renewAt - now;
+
+		// set the timeout
+		this.#renewAtTimeoutId = window.setTimeout(() => {
+			// if the document is hidden, then we need to wait for the visibility change event before renewing the token
+			if (window.document.hidden) {
+				// if the document is hidden, then wait for the visibility change event
+				window.document.addEventListener(
+					'visibilitychange',
+					() => {
+						if (!window.document.hidden) {
```

Comment:
in [DCR](https://github.com/guardian/dotcom-rendering/blob/59a8452ef70250e6256f55255752229027b01894/dotcom-rendering/src/components/Liveness.importable.tsx#L272) and [commercial](https://github.com/guardian/commercial/blob/e556b3273771a0b5d672792d7e55d348282f86c3/src/core/send-commercial-metrics.ts#L158) we use [`document.visibilityState`](https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilityState)

Reply:
updated to use `document.visibilityState` which is easier to read!